### ISSUE TITLE
Not able to re-run failed tests when same tests are being run multiple times

### ISSUE THREAD

**Issue by tarpanMpatel** [cbeust/testng#1814]
### TestNG Version
> 6.14.3

### Expected behavior
Re-run should work fine when running the same test multiple time.

### Actual behavior
Not able to re-run failed tests when same tests are being run multiple times.

### Is the issue reproductible on runner?

- [ ] Shell
- [x] Maven
- [ ] Gradle
- [ ] Ant
- [x] Eclipse
- [ ] IntelliJ
- [ ] NetBeans

### Test case sample
```xml
<listeners>
			<listener class-name="com.falkonry.utilities.RetryTestListener" />
	 	</listeners>
	<test name=" 1 .">
		<classes>
			<class name="com.falkonry.dashboard.index.DashboardIndex">
				<methods>
					<include name="test1" />
				</methods>
			</class>
		</classes>
	</test>
```

**Comment by krmahadevan** [comment]
@tarpanMpatel - I am not quite sure I understand what the issue is. Can you please help add the following

1. A simple standalone test (that doesn't have any 3rd party dependencies) that can be used to reproduce the issue
2. Exception details (if any)
3. Expected output
4. Actual output.

Without this, its hard to debug or tell you what the problem is.

**Comment by tarpanMpatel** [comment]
@krmahadevan 
 Below is my test script structure.
So What I am doing is: 
I have intentionally Failed below test "test1()".
// And Running this single tests three time parallelly. 
// my test.xml File Which I am running through the TestNG.....
// Running Parallel with thread count = 2.


**Expected output**
> If there are same tests and this same test is failed ear;y and again it comes to run and it failed then It should re-run the tests. No matter it is already re-run.
For E.x.:
my test name is: *testRerun*  (Failing)
And I am executing this test three times in one .xml (as mentioned in test.xml), Then when first time it Failed it is rerunning the test. But when the same test again failed then It should be re-run.
So, All three are failing then total 6 times my tests should run. (each one should be run 2 times as my "maxRetryCounter = 1").

**Actual output**
But here What is happening is like:
When the first test (**testRerun**) is failed then It is re-running the tests. But the test method is same and then If this tests again executing and failed then It is not re-running.

**Output**
So, As Output:
Only 4 times my tests are running.
Total 3 same tests and The first is failed so It will re-run and remaining 2 which are also failing bu as same method It is not re-running.
--------------------------------------------------------------------
----------------------------------------------------------------------------------------------
// My JAVA class which have test Which will be failed (Intentionally) and I have put this Single test three time for executing in single build

public class Rerun {
@Test
	public void testRerun() {
		int numOfFailure = 0;

		int a = 10;
		int b = 20;		
		int c = a+b;
		
		if(c==30) {
			System.out.println("Fail");
			numOfFailure ++;
		}
		else {
			System.out.println("Pass");
		}
		
		if (numOfFailure > 0) {
			Assert.assertTrue(false);
		}		
	}
----------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------
// my test.xml File Which I am running through the TestNG.....
// Running Parallel


<?xml version="1.0" encoding="UTF-8"?>
<suite name="Test" verbose="10" parallel="tests" thread-count="2">

   		<listeners>
   			<listener class-name="com.falkonry.utilities.RetryTestListener"/>
	 	</listeners>

	<test name=" 1  test rerun">
		<classes>
			<class name="com.falkonry.dashboard.index.DashboardIndex">
				<methods>
					<include name="testRerun" />
				</methods>
			</class>
		</classes>
	</test>
	
	<test name="  2  test rerun">
		<classes>
			<class name="com.falkonry.dashboard.index.DashboardIndex">
				<methods>
					<include name="testRerun" />
				</methods>
			</class>
		</classes>
	</test>
	
	<test name=" 3  test rerun">
		<classes>
			<class name="com.falkonry.dashboard.index.DashboardIndex">
				<methods>
					<include name="testRerun" />
				</methods>
			</class>
		</classes>
	</test>

</suite>
----------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------
public class RetryTestListener implements IAnnotationTransformer
{
	/**
	 * Invoke Retry Analyzer.
	 * @param annotation
	 * @param testClass
	 * @param testConstructor
	 * @param testMethod
	 * 
	 */
	public void transform(ITestAnnotation annotation, Class testClass,Constructor testConstructor, Method testMethod) 
	{
		IRetryAnalyzer retry = annotation.getRetryAnalyzer();
		if (retry == null)	
		{
			annotation.setRetryAnalyzer(RetryTest.class);
		}
	}
}
----------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------
public class  RetryTest implements IRetryAnalyzer
{
	 public int retryCounter = 0;
	 public int maxRetryCounter = 1;

	 /**
	  * Manage Retry Counter.
	  * @param result
	  * 
	  */
	 public boolean retry(ITestResult result) 
	 {
		 if (retryCounter < maxRetryCounter) 
		 {
            System.out.println("Retrying test Execution: " + result.getName() + " with status "
                    + getResultStatusName(result.getStatus()) + " for the " + (retryCounter+1) + " time(s).");
            retryCounter++;
            return true;
	      }
	        return false;
	  }
	 
	 /**
	  * Get Result Status   
	  * @param status
	  * @return
	  */
	 
     public String getResultStatusName(int status)
     {
    	 String resultName = null;
    	 if(status==1)
    	 	 resultName = "SUCCESS";
    	 if(status==2)
    		 resultName = "FAILURE";
    	 if(status==3)
    		 resultName = "SKIP";
		 return resultName;
    }

}
------------------------------------------------------------------------------------------------------------

**Comment by tarpanMpatel** [comment]
@krmahadevan 
Please find the attached class file and xml file which are the same as i have mentioned.

[test.txt](https://github.com/cbeust/testng/files/2074875/test.txt) - .xml
[Rerun.txt](https://github.com/cbeust/testng/files/2074876/Rerun.txt) - Class files
[RetryTestListener.txt](https://github.com/cbeust/testng/files/2074878/RetryTestListener.txt)

[RetryTest.txt](https://github.com/cbeust/testng/files/2074877/RetryTest.txt)

**Comment by krmahadevan** [comment]
Cannot reproduce the problem in the `master` branch. 
Looks like this issue stands fixed in the `master` branch and will be available as part of TestNG `v7.0.0`. Ideally speaking you should be able to use TestNG `v7.0.0-SNAPSHOT`. 
But currently we are having issues in ensuring that latest snapshot version is published ( That issue is being tracked via #1804 )
